client/src/components/tables/DataTable.js

```diff
@@ -0,0 +1,206 @@
+
+import dayjs from 'dayjs';
+
+import { sortObjects } from '../../utils';
+
+export const useTableStyles = makeStyles((theme) => ({
+  rootItem: {
+    paddingLeft: theme.spacing(2),
+    paddingRight: theme.spacing(2),
+    paddingTop: theme.spacing(1),
+    paddingBottom: theme.spacing(1),
+  },
+  tableItem: {
+    paddingTop: theme.spacing(4),
+    paddingRight: theme.spacing(2),
+    paddingBottom: theme.spacing(4),
+    paddingLeft: theme.spacing(2),
+  },
+  filterLabel: {
+    color: theme.palette.gray.dark,
+    fontWeight: 700,
+  },
+  actionMenuPaper: {
+    minWidth: '220px',
+  },
+  menuItem: {
+    padding: '.75rem',
+    fontSize: '17px',
+  },
+}));
+
+/**
+ * @typedef {Object} modalComponentProps
+ * @property {() => any} onSubmit              Callback form submission.
+ * @property {function} onClose                Callback for modal close.
+ * @property {boolean} open                    Whether to display the modal.
+ * @property {Object} content                  Data to pass to the dialog, such as initial form values.
+ *
+ *
+ * @typedef {Object} buttonConfig
+ * @property {string} label                    Button inner text.
+ * @property {(row: object) => any?} callback  Callback function, passed the row that the click happened on.
+ * @property {modalComponent} modal            Modal component to bind to the button.
+ *
+ *
+ * @typedef {Object} column
+ * @property {string} id                       ID string for the column, e.g. `'start_date'`.
+ *                                             Must match a property name in the target data.
+ * @property {string} name                     Label for the column, e.g. `'Start Date'`
+ * @property {undefined|'date'|'button'} type  Type of column, used for formatting.
+ *                                             * `date` will format the column as a date, using a UTC ISO timestamp.
+ *                                             * `button` will render a button (see the `button` property).
+ *                                             * `modal` will open a model (see the `button.modal` property).
+ * @property {buttonConfig?} button            Configuration for buttons, if `type` is `'button'`.
+ *
+ *
+ * @typedef {(columns: column[] ) => Promise<Object[]> } fetchFunction
+ * @typedef {(props: modalComponentProps) => JSX.Element?} modalComponent
+ */
+
+/**
+ * Data table component, renders arbitrary columns using a provided data fetching callback.
+ * @param {object}   props
+ * @param {column[]} props.columns             Array of column configuration objects.
+ *                                             Defines the table's structure and behaviour.
+ * @param {fetchFunction} props.fetchData      Async function to fetch data for this table.
+ *                                             Must return an array of objects, with keys matching column IDs.
+ * @returns {JSX.Element}
+ */
+export default ({ columns, fetchData, data }) => {
```

Comment:
Both conventions are in use throughout the project, and default export was used to fit in with some of the other components I was looking at at the time. But yes, it might be better to make it named.

Reply:
Made it named in a new commit.